File: android/app/src/main/java/com/flutcloud/flutlink/data/WebDavApi.kt
package com.flutcloud.flutlink.data

import android.util.Log
import com.flutcloud.flutlink.data.dto.WebDavEntry
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.Credentials
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import okio.BufferedSink
import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserFactory
import java.io.File
import java.io.IOException
import java.io.InputStream
import java.io.StringReader
import java.net.URLDecoder

/** Reports `(transferred, total)` bytes during a streaming transfer. */
fun interface ProgressCallback {
    fun onProgress(transferred: Long, total: Long)
}

/**
 * WebDAV client for the FlutCloud file API (PROPFIND, SEARCH, PUT, GET,
 * MKCOL, DELETE, MOVE). Mirrors the desktop `nextcloud/webdav.rs`.
 */
class WebDavApi(private val client: OkHttpClient) {

    private fun logI(msg: String) = Log.i(TAG, msg)

    private fun davRoot(session: AuthSession): String =
        "${session.normalizedBaseUrl}/remote.php/dav/files/${encodeSegment(session.username)}"

    private fun davUrl(session: AuthSession, path: String): String {
        val root = davRoot(session)
        return if (path.isEmpty() || path == "/") root else "$root/${encodePathSegments(path)}"
    }

    private fun auth(request: Request.Builder, session: AuthSession): Request.Builder =
        request.header("Authorization", Credentials.basic(session.username, session.token))

    private fun executeWebDav(request: Request, basePath: String): List<WebDavEntry> {
        val started = System.currentTimeMillis()
        try {
            client.newCall(request).execute().use { response ->
                val body = response.body?.string().orEmpty()
                if (response.code >= 400 && response.code != 207) {
                    throw ApiException(
                        "Server answered ${response.code}: $body".trim(),
                        "http_${response.code}",
                        response.code
                    )
                }
                val entries = parseMultistatus(body, basePath)
                Log.i(TAG, "${request.method} ${request.url} -> ${response.code} in ${System.currentTimeMillis() - started}ms body=${body.length} entries=${entries.size}")
                return entries
            }
        } catch (e: ApiException) {
            Log.w(TAG, "${request.method} ${request.url} -> ApiException ${e.message}", e)
            throw e
        } catch (e: IOException) {
            Log.w(TAG, "${request.method} ${request.url} -> IOException ${e.message}", e)
            throw NetworkException(e)
        }
    }

    /** List a folder (PROPFIND, Depth 1). */
    suspend fun list(session: AuthSession, path: String): List<WebDavEntry> = withContext(Dispatchers.IO) {
        val request = auth(
            Request.Builder().url(davUrl(session, path)).method("PROPFIND", null),
            session
        ).header("Depth", "1").build()
        executeWebDav(request, "/remote.php/dav/files/${encodeSegment(session.username)}")
            .filterNot { it.path == listingCurrentPath(path) }
    }

    /** WebDAV-SEARCH across the whole files tree. */
    suspend fun search(session: AuthSession, query: String): List<WebDavEntry> = withContext(Dispatchers.IO) {
        val body = searchRequestBody(session.username, query)
        val request = auth(
            Request.Builder()
                .url("${session.normalizedBaseUrl}/remote.php/dav/")
                .method("SEARCH", body.toRequestBody("application/xml".toMediaType())),
            session
        ).header("Depth", "0").build()
        executeWebDav(request, "/remote.php/dav/files/${encodeSegment(session.username)}")
    }

    /** PROPFIND (Depth 0): does the remote resource exist? */
    suspend fun exists(session: AuthSession, path: String): Boolean = withContext(Dispatchers.IO) {
        val request = auth(
            Request.Builder().url(davUrl(session, path)).method("PROPFIND", null),
            session
        ).header("Depth", "0").build()
        try {
            client.newCall(request).execute().use { response ->
                when {
                    response.code in 200..299 || response.code == 207 -> true
                    response.code == 404 -> false
                    else -> throw ApiException("Server answered ${response.code}.", "http_${response.code}", response.code)
                }
            }
        } catch (e: ApiException) {
            throw e
        } catch (e: IOException) {
            throw NetworkException(e)
        }
    }

    /** Upload bytes via PUT. Sends X-OC-MTime so the stored mtime stays stable. */
    suspend fun upload(
        session: AuthSession,
        path: String,
        bytes: ByteArray,
        contentType: String = "application/octet-stream",
        mtimeEpochSeconds: Long? = null
    ) = withContext(Dispatchers.IO) {
        val builder = auth(
            Request.Builder().url(davUrl(session, path)),
            session
        ).put(bytes.toRequestBody(contentType.toMediaType()))
        mtimeEpochSeconds?.let { builder.header("X-OC-MTime", it.toString()) }
        statusCheck(builder.build())
    }

    /**
     * Download a remote file streaming it directly into `dest` (OkHttp
     * `byteStream()` -> `FileOutputStream`), so large files never sit in the
     * heap as a `ByteArray`.
     */
    suspend fun downloadToFile(
        session: AuthSession,
        path: String,
        dest: File,
        onProgress: ProgressCallback? = null
    ): File = withContext(Dispatchers.IO) {
        val request = auth(Request.Builder().url(davUrl(session, path)).get(), session).build()
        try {
            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    throw ApiException("Download failed: HTTP ${response.code}", "http_${response.code}", response.code)
                }
                val body = response.body ?: throw ApiException("Empty download")
                val total = body.contentLength().coerceAtLeast(0L)
                dest.parentFile?.mkdirs()
                body.byteStream().use { input ->
                    dest.outputStream().use { output ->
                        val buffer = ByteArray(STREAM_BUFFER_BYTES)
                        var transferred = 0L
                        while (true) {
                            val read = input.read(buffer)
                            if (read == -1) break
                            output.write(buffer, 0, read)
                            transferred += read
                            onProgress?.onProgress(transferred, total)
                        }
                    }
                }
                dest
            }
        } catch (e: ApiException) {
            throw e
        } catch (e: IOException) {
            throw NetworkException(e)
        }
    }

    /**
     * Upload a file by streaming `openStream` (e.g. a SAF content URI) into the
     * request body, so the file never has to be read into the heap as a whole.
     * `contentLength` must be known (query the provider); for providers without
     * a size use [upload] instead.
     */
    suspend fun uploadStream(
        session: AuthSession,
        path: String,
        openStream: () -> InputStream,
        contentLength: Long,
        contentType: String = "application/octet-stream",
        mtimeEpochSeconds: Long? = null,
        onProgress: ProgressCallback? = null
    ) = withContext(Dispatchers.IO) {
        val body = object : RequestBody() {
            override fun contentType() = contentType.toMediaType()
            override fun contentLength(): Long = contentLength
            override fun writeTo(sink: BufferedSink) {
                val input = openStream() ?: throw IOException("Cannot open upload stream")
                input.use {
                    val buffer = ByteArray(STREAM_BUFFER_BYTES)
                    var transferred = 0L
                    while (true) {
                        val read = input.read(buffer)
                        if (read == -1) break
                        sink.write(buffer, 0, read)
                        transferred += read
                        onProgress?.onProgress(transferred, contentLength)
                    }
                }
            }
        }
        val builder = auth(
            Request.Builder().url(davUrl(session, path)),
            session
        ).put(body)
        mtimeEpochSeconds?.let { builder.header("X-OC-MTime", it.toString()) }
        statusCheck(builder.build())
    }

    /** Create a folder (MKCOL). 405 = already exists, treated as success. */
    suspend fun mkdir(session: AuthSession, path: String) = withContext(Dispatchers.IO) {
        val request = auth(
            Request.Builder().url(davUrl(session, path)).method("MKCOL", null),
            session
        ).build()
        statusCheck(request, ignoreStatus = 405)
    }

    /** Delete a file/folder. 404 = already gone, treated as success. */
    suspend fun delete(session: AuthSession, path: String) = withContext(Dispatchers.IO) {
        val request = auth(
            Request.Builder().url(davUrl(session, path)).method("DELETE", null),
            session
        ).build()
        statusCheck(request, ignoreStatus = 404)
    }

    /**
     * Rename/move a resource (MOVE with Overwrite: F). Throws a
     * `target_exists` ApiException when the destination already exists.
     */
    suspend fun rename(session: AuthSession, path: String, newPath: String) = withContext(Dispatchers.IO) {
        val request = auth(
            Request.Builder()
                .url(davUrl(session, path))
                .method("MOVE", null)
                .header("Destination", davUrl(session, newPath))
                .header("Overwrite", "F"),
            session
        ).build()
        try {
            client.newCall(request).execute().use { response ->
                if (response.code == 412) {
                    throw ApiException("Destination already exists: $newPath", "target_exists", 412)
                }
                if (!response.isSuccessful && response.code != 404) {
                    throw ApiException("Rename failed: HTTP ${response.code}", "http_${response.code}", response.code)
                }
            }
        } catch (e: ApiException) {
            throw e
        } catch (e: IOException) {
            throw NetworkException(e)
        }
    }

    /** Fetch a preview thumbnail; null when the server has none. */
    suspend fun preview(session: AuthSession, path: String, size: Int = 256): ByteArray? = withContext(Dispatchers.IO) {
        val url = "${session.normalizedBaseUrl}/index.php/core/preview.png?file=${encodeSegment(path)}&x=$size&y=$size"
        val request = auth(Request.Builder().url(url).get(), session).build()
        try {
            client.newCall(request).execute().use { response ->
                if (response.code == 404 || response.code == 400) return@withContext null
                if (!response.isSuccessful) {
                    throw ApiException("Preview failed: HTTP ${response.code}", "http_${response.code}", response.code)
                }
                response.body?.bytes()
            }
        } catch (e: ApiException) {
            throw e
        } catch (e: IOException) {
            throw NetworkException(e)
        }
    }

    private fun statusCheck(request: Request, ignoreStatus: Int? = null) {
        try {
            client.newCall(request).execute().use { response ->
                if (response.code == ignoreStatus) return
                if (!response.isSuccessful) {
                    throw ApiException(
                        "Server answered ${response.code}: ${response.body?.string().orEmpty()}".trim(),
                        "http_${response.code}",
                        response.code
                    )
                }
            }
        } catch (e: ApiException) {
            throw e
        } catch (e: IOException) {
            throw NetworkException(e)
        }
    }

    companion object {
        private const val TAG = "FlutLinkDav"

        /** Chunk size for streaming transfers (download + upload). */
        private const val STREAM_BUFFER_BYTES = 64 * 1024
        private fun searchRequestBody(user: String, query: String): String =
            """<?xml version="1.0" encoding="UTF-8"?>
<d:searchrequest xmlns:d="DAV:" xmlns:oc="http://owncloud.org/ns">
  <d:basicsearch>
    <d:select>
      <d:prop>
        <d:displayname/>
        <d:getcontentlength/>
        <d:getlastmodified/>
        <d:getetag/>
        <d:getcontenttype/>
      </d:prop>
    </d:select>
    <d:from>
      <d:scope>
        <d:href>/files/${escapeXml(user)}</d:href>
        <d:depth>infinity</d:depth>
      </d:scope>
    </d:from>
    <d:where>
      <d:eq>
        <d:prop><d:displayname/></d:prop>
        <d:literal>${escapeXml(query)}</d:literal>
      </d:eq>
    </d:where>
    <d:orderby/>
  </d:basicsearch>
</d:searchrequest>
"""

        private fun escapeXml(text: String): String =
            text.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")

        /** Percent-decode a segment, keeping literal `+` intact. */
        private fun urlDecode(segment: String): String =
            try {
                URLDecoder.decode(segment.replace("+", "%2B"), "UTF-8")
            } catch (_: IllegalArgumentException) {
                segment
            }

        /**
         * Normalize a client-supplied folder path into the logical relative
         * path used in listings ("" → "/", "/Photos" stays as is, trailing
         * slashes stripped). The Depth-1 response of a folder always contains
         * the folder itself, which must not appear as an entry inside its own
         * listing — otherwise clicking it is a no-op.
         */
        private fun listingCurrentPath(path: String): String {
            val trimmed = path.trim('/')
            return if (trimmed.isEmpty()) "/" else "/$trimmed"
        }

        /** Strip scheme + host from an absolute href. */
        private fun hrefPath(href: String): String {
            val idx = href.indexOf("://")
            if (idx < 0) return href
            val after = href.substring(idx + 3)
            val slash = after.indexOf('/')
            return if (slash >= 0) after.substring(slash) else "/"
        }

        /** Index right after the first `basePath` occurrence on a path boundary. */
        private fun findBasePath(path: String, basePath: String): Int? {
            var start = 0
            while (true) {
                val idx = path.indexOf(basePath, start)
                if (idx < 0) return null
                val end = idx + basePath.length
                if (end >= path.length || path[end] == '/') return end
                start = idx + 1
            }
        }

        private fun relativePath(href: String, basePath: String): String {
            val path = hrefPath(href)
            val after = findBasePath(path, basePath)?.let { path.substring(it) } ?: path
            val trimmed = after.trim('/')
            if (trimmed.isEmpty()) return "/"
            return "/" + trimmed.split('/').joinToString("/") { urlDecode(it) }
        }

        /** Flag entries under `resources` (read-only) or `parts` (write-enabled). */
        private fun classify(rel: String): Pair<Boolean, Boolean> {
            var isResource = false
            var isPart = false
            for (segment in rel.split('/')) {
                when (segment.lowercase()) {
                    "resources" -> isResource = true
                    "parts" -> isPart = true
                }
            }
            return isResource to isPart
        }

        /** Resolve a virtual link to its counterpart (`resources/<n>` ↔ `parts/<n>`). */
        private fun resolveLinkTarget(rel: String): String? {
            val segments = rel.trim('/').split('/')
            if (segments.size < 2) return null
            val target = when (segments[0].lowercase()) {
                "resources" -> "parts"
                "parts" -> "resources"
                else -> return null
            }
            return "/" + (listOf(target) + segments.drop(1)).joinToString("/")
        }

        /** Counterpart path in the paired namespace. */
        private fun pairedPath(rel: String): String? {
            val segments = rel.split('/')
            for (i in segments.indices) {
                val paired = when (segments[i].lowercase()) {
                    "resources" -> "parts"
                    "parts" -> "resources"
                    else -> continue
                }
                val copy = segments.toMutableList()
                copy[i] = paired
                return copy.joinToString("/")
            }
            return null
        }

        private fun toEntry(
            href: String,
            basePath: String,
            isDir: Boolean,
            size: Long?,
            mtime: String?,
            etag: String?,
            contentType: String?
        ): WebDavEntry? {
            val rel = relativePath(href, basePath)
            if (rel == "/" || rel.isEmpty()) return null
            val name = urlDecode(rel.substringAfterLast('/'))
            if (name.isEmpty()) return null
            val (isResource, isPart) = classify(rel)
            return WebDavEntry(
                name = name,
                path = rel,
                isDir = isDir,
                size = size,
                mtime = mtime,
                etag = etag,
                contentType = contentType,
                isResource = isResource,
                isPart = isPart,
                linkTarget = resolveLinkTarget(rel),
                pairedPath = pairedPath(rel)
            )
        }

        /** Parse a WebDAV multistatus document into structured entries. */
        fun parseMultistatus(body: String, basePath: String): List<WebDavEntry> {
            val parser = XmlPullParserFactory.newInstance().newPullParser()
            parser.setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, true)
            parser.setInput(StringReader(body))
            val entries = mutableListOf<WebDavEntry>()

            var href: String? = null
            var isDir = false
            var inResourceType = false
            var field: String? = null
            val text = StringBuilder()
            var size: Long? = null
            var mtime: String? = null
            var etag: String? = null
            var contentType: String? = null

            var event = parser.eventType
            while (event != XmlPullParser.END_DOCUMENT) {
                when (event) {
                    XmlPullParser.START_TAG -> {
                        when (parser.name) {
                            "response" -> {
                                href = null; isDir = false; size = null
                                mtime = null; etag = null; contentType = null
                            }
                            "href" -> field = "href"
                            "resourcetype" -> inResourceType = true
                            "collection" -> if (inResourceType) isDir = true
                            "getcontentlength" -> field = "size"
                            "getlastmodified" -> field = "mtime"
                            "getetag" -> field = "etag"
                            "getcontenttype" -> field = "contenttype"
                        }
                        text.setLength(0)
                    }
                    XmlPullParser.TEXT -> if (field != null) text.append(parser.text)
                    XmlPullParser.END_TAG -> {
                        if (field != null) {
                            val value = text.toString().trim()
                            when (field) {
                                "href" -> href = value
                                "size" -> size = value.toLongOrNull()
                                "mtime" -> mtime = value
                                "etag" -> etag = value
                                "contenttype" -> contentType = value
                            }
                            field = null
                        }
                        when (parser.name) {
                            "resourcetype" -> inResourceType = false
                            "response" -> {
                                href?.let {
                                    toEntry(it, basePath, isDir, size, mtime, etag, contentType)
                                        ?.let { e -> entries.add(e) }
                                }
                            }
                        }
                    }
                }
                event = parser.next()
            }
            return entries
        }
    }
}

File: ios/FlutLink/Data/Networking/WebDavApi.swift
// TEST — iOS port created by opencode. Not a production build.
import Foundation

final class WebDavApi {
    private let session: AuthSession

    init(session: AuthSession) {
        self.session = session
    }

    func withSession(_ session: AuthSession) -> WebDavApi {
        WebDavApi(session: session)
    }

    private func effectiveUser(_ targetUser: String?) -> String {
        targetUser ?? session.username
    }

    private func davRoot(_ targetUser: String? = nil) -> String {
        "\(session.normalizedBaseUrl)/remote.php/dav/files/\(effectiveUser(targetUser).urlEncoded)"
    }

    private func davUrl(path: String, targetUser: String? = nil) -> String {
        let root = davRoot(targetUser)
        if path.isEmpty || path == "/" { return root }
        return "\(root)/\(path.split(separator: "/").map { String($0).urlEncoded }.joined(separator: "/"))"
    }

    // MARK: - PROPFIND

    func list(path: String, targetUser: String? = nil) async throws -> [WebDavEntry] {
        let url = URL(string: davUrl(path: path, targetUser: targetUser))!
        var request = HttpClient.davRequest(url: url, method: "PROPFIND", session: session, targetUser: targetUser)
        request.setValue("1", forHTTPHeaderField: "Depth")
        let (data, response) = try await HttpClient.perform(request)
        guard response.statusCode == 207 || (200...299).contains(response.statusCode) else {
            let body = String(data: data, encoding: .utf8) ?? ""
            throw ApiException.api(message: "Server answered \(response.statusCode): \(body)", code: "http_\(response.statusCode)", statusCode: response.statusCode)
        }
        let basePath = "/remote.php/dav/files/\(effectiveUser(targetUser).urlEncoded)"
        let entries = Self.parseMultistatus(data: data, basePath: basePath)
        let normalizedPath = path.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let currentPath = normalizedPath.isEmpty ? "/" : "/\(normalizedPath)"
        return entries.filter { $0.path != currentPath }
    }

    // MARK: - SEARCH

    func search(query: String, targetUser: String? = nil) async throws -> [WebDavEntry] {
        let url = URL(string: "\(session.normalizedBaseUrl)/remote.php/dav/")!
        let body = Self.searchRequestBody(user: effectiveUser(targetUser), query: query)
        var request = HttpClient.davRequest(url: url, method: "SEARCH", session: session, targetUser: targetUser)
        request.setValue("0", forHTTPHeaderField: "Depth")
        request.setValue("application/xml", forHTTPHeaderField: "Content-Type")
        request.httpBody = body.data(using: .utf8)
        let (data, response) = try await HttpClient.perform(request)
        guard response.statusCode == 207 || (200...299).contains(response.statusCode) else {
            let body = String(data: data, encoding: .utf8) ?? ""
            throw ApiException.api(message: "Server answered \(response.statusCode): \(body)", code: "http_\(response.statusCode)", statusCode: response.statusCode)
        }
        let basePath = "/remote.php/dav/files/\(effectiveUser(targetUser).urlEncoded)"
        return Self.parseMultistatus(data: data, basePath: basePath)
    }

    // MARK: - Exists

    func exists(path: String, targetUser: String? = nil) async throws -> Bool {
        let url = URL(string: davUrl(path: path, targetUser: targetUser))!
        var request = HttpClient.davRequest(url: url, method: "PROPFIND", session: session, targetUser: targetUser)
        request.setValue("0", forHTTPHeaderField: "Depth")
        let (_, response) = try await HttpClient.perform(request)
        switch response.statusCode {
        case 200...299, 207: return true
        case 404: return false
        default:
            throw ApiException.api(message: "Server answered \(response.statusCode).", code: "http_\(response.statusCode)", statusCode: response.statusCode)
        }
    }

    // MARK: - Upload (PUT)

    func upload(path: String, data: Data, contentType: String = "application/octet-stream", mtimeEpochSeconds: Int64? = nil, targetUser: String? = nil) async throws {
        let url = URL(string: davUrl(path: path, targetUser: targetUser))!
        var request = HttpClient.davRequest(url: url, method: "PUT", session: session, targetUser: targetUser)
        request.httpBody = data
        request.setValue(contentType, forHTTPHeaderField: "Content-Type")
        if let mtime = mtimeEpochSeconds { request.setValue("\(mtime)", forHTTPHeaderField: "X-OC-MTime") }
        try await performStatusCheck(request)
    }

    func uploadStream(
        path: String,
        data: Data,
        contentType: String = "application/octet-stream",
        mtimeEpochSeconds: Int64? = nil,
        targetUser: String? = nil,
        onProgress: ((Int64, Int64) -> Void)? = nil
    ) async throws {
        try await upload(path: path, data: data, contentType: contentType, mtimeEpochSeconds: mtimeEpochSeconds, targetUser: targetUser)
        onProgress?(Int64(data.count), Int64(data.count))
    }

    // MARK: - Download (GET)

    func downloadToFile(path: String, targetUser: String? = nil, onProgress: ((Int64, Int64) -> Void)? = nil) async throws -> Data {
        let url = URL(string: davUrl(path: path, targetUser: targetUser))!
        let request = HttpClient.davRequest(url: url, method: "GET", session: session, targetUser: targetUser)
        let (tempURL, response) = try await URLSession.shared.download(for: request)
        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
            let statusCode = (response as? HTTPURLResponse)?.statusCode ?? -1
            throw ApiException.api(message: "Download failed: HTTP \(statusCode)", code: "http_\(statusCode)", statusCode: statusCode)
        }
        let data = try Data(contentsOf: tempURL)
        let total = Int64(data.count)
        onProgress?(total, total)
        return data
    }

    // MARK: - MKCOL

    func mkdir(path: String, targetUser: String? = nil) async throws {
        let url = URL(string: davUrl(path: path, targetUser: targetUser))!
        let request = HttpClient.davRequest(url: url, method: "MKCOL", session: session, targetUser: targetUser)
        try await performStatusCheck(request, ignoreStatus: 405)
    }

    // MARK: - DELETE

    func delete(path: String, targetUser: String? = nil) async throws {
        let url = URL(string: davUrl(path: path, targetUser: targetUser))!
        let request = HttpClient.davRequest(url: url, method: "DELETE", session: session, targetUser: targetUser)
        try await performStatusCheck(request, ignoreStatus: 404)
    }

    // MARK: - MOVE (rename)

    func rename(path: String, newPath: String, targetUser: String? = nil) async throws {
        let url = URL(string: davUrl(path: path, targetUser: targetUser))!
        var request = HttpClient.davRequest(url: url, method: "MOVE", session: session, targetUser: targetUser)
        request.setValue(davUrl(path: newPath, targetUser: targetUser), forHTTPHeaderField: "Destination")
        request.setValue("F", forHTTPHeaderField: "Overwrite")
        let (_, response) = try await HttpClient.perform(request)
        if response.statusCode == 412 {
            throw ApiException.api(message: "Destination already exists: \(newPath)", code: "target_exists", statusCode: 412)
        }
        guard response.statusCode == 200 || response.statusCode == 201 || response.statusCode == 204 || response.statusCode == 404 else {
            throw ApiException.api(message: "Rename failed: HTTP \(response.statusCode)", code: "http_\(response.statusCode)", statusCode: response.statusCode)
        }
    }

    // MARK: - Helpers

    private func performStatusCheck(_ request: URLRequest, ignoreStatus: Int? = nil) async throws {
        let (_, response) = try await HttpClient.perform(request)
        if let ignore = ignoreStatus, response.statusCode == ignore { return }
        guard (200...299).contains(response.statusCode) else {
            let body = String(data: Data(), encoding: .utf8) ?? ""
            throw ApiException.api(message: "Server answered \(response.statusCode): \(body)", code: "http_\(response.statusCode)", statusCode: response.statusCode)
        }
    }

    // MARK: - XML Parsing

    static func parseMultistatus(data: Data, basePath: String) -> [WebDavEntry] {
        let parser = WebDavXMLParser(data: data, basePath: basePath)
        return parser.parse()
    }

    private static func searchRequestBody(user: String, query: String) -> String {
        """
        <?xml version="1.0" encoding="UTF-8"?>
        <d:searchrequest xmlns:d="DAV:" xmlns:oc="http://owncloud.org/ns">
          <d:basicsearch>
            <d:select>
              <d:prop>
                <d:displayname/>
                <d:getcontentlength/>
                <d:getlastmodified/>
                <d:getetag/>
                <d:getcontenttype/>
              </d:prop>
            </d:select>
            <d:from>
              <d:scope>
                <d:href>/files/\(escapeXml(user))</d:href>
                <d:depth>infinity</d:depth>
              </d:scope>
            </d:from>
            <d:where>
              <d:eq>
                <d:prop><d:displayname/></d:prop>
                <d:literal>\(escapeXml(query))</d:literal>
              </d:eq>
            </d:where>
            <d:orderby/>
          </d:basicsearch>
        </d:searchrequest>
        """
    }

    private static func escapeXml(_ text: String) -> String {
        text.replacingOccurrences(of: "&", with: "&amp;")
            .replacingOccurrences(of: "<", with: "&lt;")
            .replacingOccurrences(of: ">", with: "&gt;")
    }

    static func hrefPath(_ href: String) -> String {
        guard let range = href.range(of: "://") else { return href }
        let after = String(href[range.upperBound...])
        guard let slashIdx = after.firstIndex(of: "/") else { return "/" }
        return String(after[slashIdx...])
    }

    static func relativePath(href: String, basePath: String) -> String {
        let path = hrefPath(href)
        guard let range = path.range(of: basePath) else {
            let trimmed = path.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
            return trimmed.isEmpty ? "/" : "/" + trimmed
        }
        let after = String(path[range.upperBound...])
        let trimmed = after.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        if trimmed.isEmpty { return "/" }
        return "/" + trimmed.split(separator: "/").map {
            $0.removingPercentEncoding ?? String($0)
        }.joined(separator: "/")
    }

    static func classify(_ rel: String) -> (isResource: Bool, isPart: Bool) {
        var isResource = false
        var isPart = false
        for segment in rel.split(separator: "/") {
            switch segment.lowercased() {
            case "resources": isResource = true
            case "parts": isPart = true
            default: break
            }
        }
        return (isResource, isPart)
    }

    static func resolveLinkTarget(_ rel: String) -> String? {
        let segments = rel.trimmingCharacters(in: CharacterSet(charactersIn: "/")).split(separator: "/")
        guard segments.count >= 2 else { return nil }
        let target: String
        switch segments[0].lowercased() {
        case "resources": target = "parts"
        case "parts": target = "resources"
        default: return nil
        }
        let tail = segments.dropFirst(1).map { s in String(s) }
        return "/" + ([target] + tail).joined(separator: "/")
    }

    static func pairedPath(_ rel: String) -> String? {
        let segments = rel.split(separator: "/")
        for (i, segment) in segments.enumerated() {
            let paired: String
            switch segment.lowercased() {
            case "resources": paired = "parts"
            case "parts": paired = "resources"
            default: continue
            }
            var copy = segments.map(String.init)
            copy[i] = paired
            return copy.joined(separator: "/")
        }
        return nil
    }
}

// MARK: - XML Parser

final class WebDavXMLParser: NSObject, XMLParserDelegate {
    private var entries: [WebDavEntry] = []
    private let basePath: String

    private var currentHref: String?
    private var currentIsDir = false
    private var inResourceType = false
    private var currentField: String?
    private var textBuffer = ""
    private var currentSize: Int64?
    private var currentMtime: String?
    private var currentEtag: String?
    private var currentContentType: String?

    init(data: Data, basePath: String) {
        self.basePath = basePath
        super.init()
        let parser = XMLParser(data: data)
        parser.delegate = self
        parser.parse()
    }

    func parse() -> [WebDavEntry] {
        return entries
    }

    func parser(_ parser: XMLParser, didStartElement elementName: String, namespaceURI: String?, qualifiedName qName: String?, attributes attributeDict: [String: String] = [:]) {
        switch elementName {
        case "response":
            currentHref = nil; currentIsDir = false; currentSize = nil
            currentMtime = nil; currentEtag = nil; currentContentType = nil
        case "href": currentField = "href"
        case "resourcetype": inResourceType = true
        case "collection": if inResourceType { currentIsDir = true }
        case "getcontentlength": currentField = "size"
        case "getlastmodified": currentField = "mtime"
        case "getetag": currentField = "etag"
        case "getcontenttype": currentField = "contenttype"
        default: break
        }
        textBuffer = ""
    }

    func parser(_ parser: XMLParser, foundCharacters string: String) {
        if currentField != nil { textBuffer += string }
    }

    func parser(_ parser: XMLParser, didEndElement elementName: String, namespaceURI: String?, qualifiedName qName: String?) {
        if let field = currentField {
            let value = textBuffer.trimmingCharacters(in: .whitespacesAndNewlines)
            switch field {
            case "href": currentHref = value
            case "size": currentSize = Int64(value)
            case "mtime": currentMtime = value
            case "etag": currentEtag = value
            case "contenttype": currentContentType = value
            default: break
            }
            currentField = nil
        }
        switch elementName {
        case "resourcetype": inResourceType = false
        case "response":
            if let href = currentHref {
                let rel = WebDavApi.relativePath(href: href, basePath: basePath)
                if rel != "/" && !rel.isEmpty {
                    let name = String(rel.split(separator: "/").last ?? Substring(rel))
                    if !name.isEmpty {
                        let (isResource, isPart) = WebDavApi.classify(rel)
                        entries.append(WebDavEntry(
                            name: name,
                            path: rel,
                            isDir: currentIsDir,
                            size: currentSize,
                            mtime: currentMtime,
                            etag: currentEtag,
                            contentType: currentContentType,
                            isResource: isResource,
                            isPart: isPart,
                            linkTarget: WebDavApi.resolveLinkTarget(rel),
                            pairedPath: WebDavApi.pairedPath(rel)
                        ))
                    }
                }
            }
        default: break
        }
    }
}
